#### Project Type: Data Wrangling
### Project Number: 0005
## Project Title: Data Scientist Job Market in the U.S

---

Источник данных: https://www.kaggle.com/sl6149/data-scientist-job-market-in-the-us?select=fulltimeAL.csv


# **STAGE 1. PROBLEM DEFINITION**
    




### Project Objectives

---

1. Развитие навыка очистки и подготовки данных для проведения дальнейших исследований 
2. Дааные после проекта должны быть более "высокого" качества, чем до. 
3. Очередной проект в портфолио 
4. Удавольствие от работы с данными 

### Success criteria

---

1. Отсувие пропусков 
2. Отсувие дублирующих значений 
3. Данные должны быть приведены к форме принятой в анализе 
4. "Чистые" данные

### Task type

---

Data Wrangling

    
# **STAGE 2. PRELIMINARY DATA WRANGLING**



### Import Libraries

---

In [54]:
# Data manipulation
import pandas as pd
import numpy as np

# Data Visualizations
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

### Import Data

---

Исходный набор данных содержит 16 файлов для анализа. Прежде чем загружать их все исследуем несколько из них.
- Один файл называется alldata 
- Оставшиеся называются fulltime + сокращение заглавными буквами название города или региона 

In [55]:
# Обобщенный путь к данным 
data_path = r'D:/Data_Science_Project/1. Data Wrangling/[0005] Project 3 - Data Scientist Job Market in the U.S'

## SECTION 2.1.  DATA UNDERSTANDING

### _Collect initial data_

---

### Обзор файла alldata

In [56]:
# Загрузка первого файла в наборе
alldata =  pd.read_csv(data_path + '/alldata.csv')
alldata.head()

position             company  \
0                               Development Director             ALS TDI   
1  An Ostentatiously-Excitable Principal Research...  The Hexagon Lavish   
2                                     Data Scientist      Xpert Staffing   
3                                       Data Analyst      Operation HOPE   
4  Assistant Professor -TT - Signal Processing & ...    Emory University   

                                         description  reviews  \
0  Development Director\nALS Therapy Development ...      NaN   
1  Job Description\n\n"The road that leads to acc...      NaN   
2  Growing company located in the Atlanta, GA are...      NaN   
3  DEPARTMENT: Program OperationsPOSITION LOCATIO...     44.0   
4  DESCRIPTION\nThe Emory University Department o...    550.0   

             location  
0  Atlanta, GA 30301   
1         Atlanta, GA  
2         Atlanta, GA  
3  Atlanta, GA 30303   
4         Atlanta, GA

In [57]:
alldata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6964 entries, 0 to 6963
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   position     6953 non-null   object 
 1   company      6953 non-null   object 
 2   description  6953 non-null   object 
 3   reviews      5326 non-null   float64
 4   location     6953 non-null   object 
dtypes: float64(1), object(4)
memory usage: 272.2+ KB


In [58]:
alldata.tail().T

6959  \
position             Data Developer / Machine Learning Analyst   
company                                                 NetApp   
description  Are you data-driven? We at NetApp believe in t...   
reviews                                                    574   
location                                         Sunnyvale, CA   

                                                          6960  \
position                                           Scientist I   
company                       Pharmacyclics, an Abbvie Company   
description  Pharmacyclics is committed to the development ...   
reviews                                                     26   
location                                         Sunnyvale, CA   

                                                          6961  \
position                                      Intern Scientist   
company                                               Oath Inc   
description  Oath, a subsidiary of Verizon, is a values-led...   
reviews                                                      5   
location                                         Sunnyvale, CA   

                                                          6962  \
position                       Senior Data & Applied Scientist   
company                                              Microsoft   
description  We are the Bing Core Relevance team responsibl...   
reviews                                                   4618   
location                                         Sunnyvale, CA   

                                                          6963  
position               Principal Data Scientist, Deep Learning  
company                                                Comcast  
description  Comcast’s Technology &amp; Product organizatio...  
reviews                                                  11610  
location                                   Sunnyvale, CA 94089

### Обзор файлов  fulltimeAL и fulltimeNY

In [59]:
# Загрузка файла 
fulltimeAL =  pd.read_csv(data_path + '/fulltimeAL.csv')
fulltimeAL.head()

position             company  \
0                               Development Director             ALS TDI   
1  An Ostentatiously-Excitable Principal Research...  The Hexagon Lavish   
2                                     Data Scientist      Xpert Staffing   
3                                       Data Analyst      Operation HOPE   
4  Assistant Professor -TT - Signal Processing & ...    Emory University   

                                         description    numreview  \
0  <span id="job_summary" class="summary"><div><d...          NaN   
1  <span id="job_summary" class="summary"><div>Jo...          NaN   
2  <span id="job_summary" class="summary"><div><d...          NaN   
3  <span id="job_summary" class="summary"><p><b>D...   44 reviews   
4  <span id="job_summary" class="summary"><div><h...  550 reviews   

             location  
0  Atlanta, GA 30301   
1         Atlanta, GA  
2         Atlanta, GA  
3  Atlanta, GA 30303   
4         Atlanta, GA

In [60]:
fulltimeAL.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 269 entries, 0 to 268
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   position     269 non-null    object
 1   company      269 non-null    object
 2   description  269 non-null    object
 3   numreview    227 non-null    object
 4   location     269 non-null    object
dtypes: object(5)
memory usage: 10.6+ KB


In [61]:
# Загрузка файла 
fulltimeNY =  pd.read_csv(data_path + '/fulltimeNY.csv')
fulltimeNY.head()

position  \
0  Data Scientist—Research (Ref # EXEC/RD_DAT_NYC...   
1                  Data Scientist/Pre Sales Engineer   
2                    Data Analyst/Research Scientist   
3                                     Data Scientist   
4                                     Data Scientist   

                                             company  \
0  New York State Office of the Attorney General ...   
1                                            Celonis   
2                 Columbia University Medical Center   
3                                   fractalanalytics   
4                               Blackboard Insurance   

                                         description    numreview  \
0  <span id="job_summary" class="summary"><p><b>E...          NaN   
1  <span id="job_summary" class="summary"><p>Celo...          NaN   
2  <span id="job_summary" class="summary"><p>We a...  217 reviews   
3  <span id="job_summary" class="summary"><div><p...          NaN   
4  <span id="job_summary" class="summary"><div><p...          NaN   

             location  
0        New York, NY  
1        New York, NY  
2  New York, NY 10032  
3     Jersey City, NJ  
4        New York, NY

In [62]:
fulltimeNY.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 986 entries, 0 to 985
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   position     986 non-null    object
 1   company      986 non-null    object
 2   description  986 non-null    object
 3   numreview    755 non-null    object
 4   location     986 non-null    object
dtypes: object(5)
memory usage: 38.6+ KB


Анализ таблиц показывает, что все данные уже сведены в одну таблицу alldata поэтому дальнейшую работу будем вести с ней.

### _Describe data_

---

In [63]:
# Форма исходного набора данных
print("Форма исходного набора данных:\n{}".format(alldata.shape))

Форма исходного набора данных:
(6964, 5)


In [64]:
# Список столбцов
alldata.columns.tolist()

['position', 'company', 'description', 'reviews', 'location']

In [65]:
# Анализ типов данных
alldata.dtypes 

position        object
company         object
description     object
reviews        float64
location        object
dtype: object

In [66]:
# Общая информация о наборе данных 
alldata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6964 entries, 0 to 6963
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   position     6953 non-null   object 
 1   company      6953 non-null   object 
 2   description  6953 non-null   object 
 3   reviews      5326 non-null   float64
 4   location     6953 non-null   object 
dtypes: float64(1), object(4)
memory usage: 272.2+ KB


#### Описание переменных набора данных

1. position - Занимаемая должность 
1. company - Название компании 
1. description - Описание вакансии 
1. reviews - колличество просмотров вакансии
1. location - местоположение офиса

#### Описание типов переменных

1. position - Категориальная 
1. company - Категориальная 
1. description - Категориальная 
1. reviews - Колличественная 
1. location - Категориальная 

## SECTION 2.2 DATA PREPARATION

### _Verify data quality_

---

In [67]:
# Функция коплексной оценки качества данных 
def verify_data_quality (evaluated_dataset):
    # Проверка типов данных
    print("Типы данных:\n{}\n".format(evaluated_dataset.dtypes) + '_' * 100+ '\n')
    # Проверка наличия значений NaN
    print("Пропущенные значения:\n{}\n".format(evaluated_dataset.isnull().sum().sum()) + '_' * 100+ '\n')
    # Количество значений, отличных от NaN
    print("Количество значений, отличных от NaN:\n{}\n".format(evaluated_dataset.count().sum()) + '_' * 100+ '\n')
    # Поиск дублирующих строк 
    print("Дублирующих строк:\n{}\n".format(evaluated_dataset.duplicated().sum()) + '_' * 100+ '\n')
    # Форма нарбора данных
    print("Форма нарбора данных:\n{}\n".format(evaluated_dataset.shape) + '_' * 100+ '\n')
    # Тип индекса набора данных
    print("Тип индекса набора данных:\n{}\n".format(evaluated_dataset.index) + '_' * 100+ '\n')

In [68]:
# Оценка качетва данных
verify_data_quality(alldata)

Типы данных:
position        object
company         object
description     object
reviews        float64
location        object
dtype: object
____________________________________________________________________________________________________

Пропущенные значения:
1682
____________________________________________________________________________________________________

Количество значений, отличных от NaN:
33138
____________________________________________________________________________________________________

Дублирующих строк:
10
____________________________________________________________________________________________________

Форма нарбора данных:
(6964, 5)
____________________________________________________________________________________________________

Тип индекса набора данных:
RangeIndex(start=0, stop=6964, step=1)
____________________________________________________________________________________________________



### _Data type conversions_

---

In [69]:
# Список столбцов
alldata.columns.tolist()

['position', 'company', 'description', 'reviews', 'location']

In [72]:
# Приведение к типу данных category
alldata[['position', 'company', 'description', 'location']] = alldata[['position', 'company', 'description',  'location']].astype('category')

# Оценка типов данныз набора
alldata.dtypes

position       category
company        category
description    category
reviews         float64
location       category
dtype: object

### _Conclusions_

---

На данной стадии подготовки данных выполнены следующие действия:  

- Общая оценка качества данных
- Установлено, что имеется:
    - 1682 пропущенных значения 
    - 10 дублирующих строк 
- Выполнена конвертация типов данных 
    - position, company, description, location: как тип category 
    - reviews: как тип float64

Дальнейшая подготовка данных выполненна в следующем разделе проекта. 

In [79]:
# Сохранение промежуточных результатов 
pd.DataFrame.to_feather(alldata, 'alldata_prepated.file')